magento/pwa-studio
# Added feature to pick from multiple sample backends.

## Description

This PR introduces a new extension and a new feature that allows developers to choose which Magento Backend they want to use while developing. 

While creating a new app, developers will be prompted to choose a Magento backend from the sample backends provided by the PWA Team or enter their own backend URL. The sample backends are fetched at runtime from a secure location hosted and maintained by the PWA Team.

This PR also introduces a new Target that accepts env validator functions. These functions will receive the project env in the JSON format and they can run validations and report errors to the buildpack's build process. The extension added in the PR uses this target to add a validator that checks the backend URL and provides options if needed.

## Related Issue
Closes [PWA-96](https://jira.corp.magento.com/browse/PWA-96)

### Verification Stakeholders
@zetlen 
@jimbo 
@dpatil-magento 

### Specification
We need to update the docs about this change. @jcalcaben can you advise where I need to update these docs.

### Verification Steps

Scaffold test
1. Checkout this branch.
2. `DEBUG_PROJECT_CREATION=1 <Path to PWA Studio Package>/packages/create-pwa/bin/create-pwa`
3. You should be able to see options from the new API.

![image](https://user-images.githubusercontent.com/35203638/98605429-085d0d00-22ab-11eb-9a16-b138da39ef1d.png)

`Magento 2.3.3 with Venia sample data installed` from the above image is fetched from the API.

Build process
1. Checkout this branch.
2. Change the `MAGENTO_BACKEND_URL` to an unreachable URL.
3. Run `yarn build`. The build process should fail and suggest some sample backends.

![image](https://user-images.githubusercontent.com/35203638/98606753-6c80d080-22ad-11eb-9869-9badd9dc563e.png)

If the backend is fine, nothing should change from how `develop` works.

## Checklist

* I have added tests to cover my changes, if necessary.
* I have added translations for new strings, if necessary.
* I have updated the documentation accordingly, if necessary.
